src/java_bytecode/java_types.cpp

```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+    if(e_pos == std::string::npos)
+      throw unsupported_java_class_signature_exceptiont("recursive generic");
+
+    // erase generic information between brackets
+    class_name.erase(f_pos, e_pos - f_pos + 1);
+
+    // Search the remainder of the string for generic signature
+    f_pos = class_name.find('<', e_pos + 1);
+  }
+  return class_name;
+}
+
+/// Returns the full class name, skipping over the generics
+/// \param src: a type descriptor or signature like LOuterClass<TT;>.Inner;
+/// \return The full name of the class like OuterClass.Inner
+std::string gather_full_class_name(const std::string &src)
+{
+  PRECONDITION(src[0] == 'L');
+  PRECONDITION(src[src.size() - 1] == ';');
+
+  std::string class_name = src.substr(1, src.size() - 2);
+
+  class_name = erase_type_arguments(class_name);
+
+  std::replace(class_name.begin(), class_name.end(), '.', '$');
+  std::replace(class_name.begin(), class_name.end(), '/', '.');
+  return class_name;
+}
+
+size_t find_closing_semi_colon_for_reference_type(
+  const std::string src,
+  size_t starting_point = 0);
+
+/// Given a substring of a descriptor or signature that contains 1 or more types
+/// parse out the inidividual types. This is used for parsing the parameters of
+/// a function or the gneeric arguments contained within angle brackets.
+/// \param src: The input string that is wrapped in either ( ) or < >
+/// \param class_name_prefix: The name of the class to use to prefix any found
+///   generic types
+/// \param opening_bracket: For checking string is passed in as expected, the
+///   opening bracket (i.e. '(' or '<').
+/// \param closing_bracket: For checking string is passed in as expected, the
+///   closing bracket (i.e. ')' or '>').
+/// \return A vector of types that the string represents
+std::vector<typet> parse_list_types(
+  const std::string src,
+  const std::string class_name_prefix,
+  const char opening_bracket,
+  const char closing_bracket)
+{
+  PRECONDITION(src[0] == opening_bracket);
+  PRECONDITION(src[src.size() - 1] == closing_bracket);
+  std::vector<typet> type_list;
+  for(std::size_t i = 1; i < src.size() - 1; i++)
+  {
+    size_t start = i;
+    while(i < src.size())
+    {
+      // parameter is an object type or instantiated generic type
+      if(src[i] == 'L')
+      {
+        i = find_closing_semi_colon_for_reference_type(src, i);
+        break;
+      }
+
+      // parameter is an array
+      else if(src[i] == '[')
+        i++;
+
+      // parameter is a type variable
+      else if(src[i] == 'T')
+        i = src.find(';', i); // ends on ;
+
+      // type is an atomic type (just one character)
+      else
+      {
+        break;
+      }
+    }
+
+    std::string sub_str = src.substr(start, i - start + 1);
+    const typet &new_type = java_type_from_string(sub_str, class_name_prefix);
+    INVARIANT(new_type != nil_typet(), "Failed to parse type");
+
+    if(new_type.id() == ID_symbol)
```

Comment:
why not an invariant with information string?

Reply:
Removed as was only there as [previously we did](https://github.com/diffblue/cbmc/pull/1512/files#diff-ba71807a8e04db80ccb196e4eb6465feL281):

```
if(param.type().id()==ID_symbol)		
  param.type()=java_reference_type(param.type());
```

but afaict this never happens (wanted to verify this with a hard fail - but since the CI passes and I can't see a way for this to happen, I guess it is just legacy code)